core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -278,55 +304,63 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
-      val logInfos: Seq[FileStatus] = statusList
-        .filter { entry =>
-          try {
-            val prevFileSize = fileToAppInfo.get(entry.getPath()).map{_.fileSize}.getOrElse(0L)
-            !entry.isDirectory() && prevFileSize < entry.getLen()
-          } catch {
-            case e: AccessControlException =>
-              // Do not use "logInfo" since these messages can get pretty noisy if printed on
-              // every poll.
-              logDebug(s"No permission to read $entry, ignoring.")
-              false
+    metrics.updateCount.inc()
+    metrics.updateLastAttempted.touch()
+    time(metrics.updateTimer) {
+      try {
+        val newLastScanTime = getNewLastScanTime()
+        logDebug(s"Scanning $logDir with lastScanTime==$lastScanTime")
+        val statusList = Option(fs.listStatus(new Path(logDir))).map(_.toSeq)
+          .getOrElse(Seq[FileStatus]())
+        // scan for modified applications, replay and merge them
+        val logInfos: Seq[FileStatus] = statusList
+          .filter { entry =>
+            try {
+              val prevFileSize = fileToAppInfo.get(entry.getPath()).map{_.fileSize}.getOrElse(0L)
+              !entry.isDirectory() && prevFileSize < entry.getLen()
+            } catch {
+              case e: AccessControlException =>
+                // Do not use "logInfo" since these messages can get pretty noisy if printed on
+                // every poll.
+                logDebug(s"No permission to read $entry, ignoring.")
+                false
+            }
           }
+          .flatMap { entry => Some(entry) }
+          .sortWith { case (entry1, entry2) =>
+            entry1.getModificationTime() >= entry2.getModificationTime()
         }
-        .flatMap { entry => Some(entry) }
-        .sortWith { case (entry1, entry2) =>
-          entry1.getModificationTime() >= entry2.getModificationTime()
-      }
 
-      if (logInfos.nonEmpty) {
-        logDebug(s"New/updated attempts found: ${logInfos.size} ${logInfos.map(_.getPath)}")
-      }
-      logInfos.map { file =>
-          replayExecutor.submit(new Runnable {
-            override def run(): Unit = mergeApplicationListing(file)
-          })
+        if (logInfos.nonEmpty) {
+          logDebug(s"New/updated attempts found: ${logInfos.size} ${logInfos.map(_.getPath)}")
         }
-        .foreach { task =>
-          try {
-            // Wait for all tasks to finish. This makes sure that checkForLogs
-            // is not scheduled again while some tasks are already running in
-            // the replayExecutor.
-            task.get()
-          } catch {
-            case e: InterruptedException =>
-              throw e
-            case e: Exception =>
-              logError("Exception while merging application listings", e)
+        logInfos.map { file =>
+            replayExecutor.submit(new Runnable {
+              override def run(): Unit = time(metrics.mergeApplicationListingTimer) {
```

Comment:
This metric contains the time to replay an app event log, which can vary a lot per application; same issue with the UI load time issue I pointed out before. That makes this metric not very useful.